provisioning/main.c

```diff
@@ -0,0 +1,321 @@
+    status = psa_generate_random(output, output_len);
+
+    /* Fail immediately if our source of randomness fails. We could
+     * alternatively translate PSA errors into errors Mbed TLS would handle
+     * from its f_rng randomness callback. */
+    MBED_ASSERT(status != PSA_SUCCESS);
+
+    return 0;
+}
+
+psa_status_t generate_key_on_device(void)
+{
+    psa_status_t status;
+    psa_key_attributes_t attributes = PSA_KEY_ATTRIBUTES_INIT;
+    psa_key_handle_t handle;
+
+    /* Set device private key attributes. */
+    /* Note: It is not necessary to hard-code the physical slot number within
+     * the secure element. The secure element driver can automatically allocate
+     * a slot that fits your use case. */
+    psa_set_key_slot_number(&attributes, EXAMPLE_GENERATED_KEY_SE_SLOT);
+    psa_set_key_id(&attributes, EXAMPLE_GENERATED_KEY_ID);
+    psa_set_key_lifetime(&attributes, PSA_ATECC608A_LIFETIME);
+    psa_set_key_algorithm(&attributes, PSA_ALG_ECDSA(PSA_ALG_SHA_256));
+    psa_set_key_bits(&attributes, 256);
+    psa_set_key_usage_flags(&attributes, PSA_KEY_USAGE_SIGN);
+    psa_set_key_type(&attributes,
+                     PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_CURVE_SECP256R1));
+
+    /* Generate the key inside the secure element. */
+    status = psa_generate_key(&attributes, &handle);
+    if (status != PSA_SUCCESS) {
+        return status;
+    }
+
+    return psa_close_key(handle);
+}
+
+psa_status_t register_preprovisioned_keys(void)
+{
+    psa_key_attributes_t attributes = PSA_KEY_ATTRIBUTES_INIT;
+
+    /* Set device private key attributes. */
+    psa_set_key_slot_number(&attributes, EXAMPLE_FACTORY_KEY_SE_SLOT);
+    psa_set_key_id(&attributes, EXAMPLE_FACTORY_KEY_ID);
+    psa_set_key_lifetime(&attributes, PSA_ATECC608A_LIFETIME);
+    psa_set_key_algorithm(&attributes, PSA_ALG_ECDSA(PSA_ALG_SHA_256));
+    psa_set_key_bits(&attributes, 256);
+    psa_set_key_usage_flags(&attributes, PSA_KEY_USAGE_SIGN);
+    psa_set_key_type(&attributes,
+                     PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_CURVE_SECP256R1));
+
+    /* Register the factory-created key with Mbed Crypto, so that Mbed Crypto
+     * knows that the key exists and how to find and access the key. This
+     * registration only needs doing once, as Mbed Crypto will remember the
+     * registration even across reboots. */
+    return mbedtls_psa_register_se_key(&attributes);
+}
+
+void print_public_key(psa_key_id_t key_id)
+{
+    enum { OUTPUT_LEN = 256 };
+    int ret;
+    psa_status_t status;
+    unsigned char *output;
+    psa_key_handle_t handle;
+    mbedtls_pk_context pk;
+
+    printf("\tKey ID %lu:\n", key_id);
+
+    /* Open the specified key. */
+    status = psa_open_key(key_id, &handle);
+    if (status != PSA_SUCCESS)
+    {
+        printf("Failed to open key %lu with status=%ld\n", key_id, status);
+        return;
+    }
+
+    output = calloc(1, OUTPUT_LEN);
+    if (!output) {
+        puts("Out of memory");
+        return;
+    }
+
```

Comment:
If you read this program expecting a demonstration of the PSA API, it's a bit weird to see mbedtls API functions here. Please add a comment stating that the PSA API currently lacks a way to format the key in the desired format, so we use the mbedtls pk module.